docs/docs/async-dialogs.md

```diff
@@ -4,3 +4,628 @@ title: Async Dialogs
+
+```javascript
+Dialog.prototype.cancel = function() {
+  this._cancelCallback(new UserCanceledError());
+};
+
+…
+
+PromptDialog.prototype.attachDomEvents = function() {
+  var _this = this;
+  this.okButton.addEventListener('click', function() {
+    _this._okCallback(_this.inputEl.value);
+  });
+  this.cancelButton.addEventListener('click', function() {
+    _this.cancel();
+  });
+};
+```
+
+And in our usage case we can test for it:
+
+```javascript
+// Timeout the dialog in five seconds.
+setTimeout(function() { prompt.cancel(); }, 5000);
+
+prompt.show('What is your name?')
+  .waitForUser()
+  .then(function(name) {
+    output.innerHTML = '' + name;
+  })
+  .catch(UserCanceledError, function() {
+    output.innerHTML = '¯\\_(ツ)_/¯';
+  })
+  .catch(function(e) {
+    console.log('Something bad happened!', e);
+  })
+  .finally(function() {
+    prompt.hide();
+  });
+```
+
+[Run example on JSBin](http://jsbin.com/yaropo/edit?js,output)
+
+**NOTE:** Bluebird supports [cancellation](api/cancellation.html) as an optional
+feature that is turned off by default. However, its implementation (since
+version 3.0) is meant to stop the then and catch callbacks from firing. It is
+not helpful in the example of a user cancellation as described here.
+
+## Progress bar
+
+When there are asynchronous tasks that have the ability to notify progress as
+they complete it can be tempting to want that in the promise that represents
+that task. Unfortunately this is a bit of an anti-pattern. That is because the
+point of promises is to represent a value as if it was natural (like it is in
+normal synchronous code) and not to be over glorified callback management.
+
+So how then could we represent a progress bar like dialog? Well the answer is to
+manage the progress through callbacks outside the promise API. Bluebird has
+since [deprecated the progression feature](deprecated-apis.html#progression) and
+offers an alternative which I hope to illustrate here.
+
+Another key difference between a *progress bar* dialog and any other dialog
+we've discussed here is that a progress bar represents information on another
+task and *not* user import. Instead of the program waiting for the user to
+provide a value the dialog box is waiting on the program to provide a value
+(resolved: 100% complete, rejected: aborted half way through). Because of this
+the *progress bar* dialog would have a different interface then the previous
+dialogs we've covered. However, there can still be some user interaction so in
+essence we are dealing with two promises.
+
+Bluebird has a way to manage more than one promise simultaneously. When you want
+to know if more then one promise completes there is a `Promise.all()` function
+that takes an array of promises and returns a new promise waiting for them all
+to resolve. But if any one is rejected the returned promise is immediately
+rejected.
+
+Bluebird also has a `Promise.race()` function which does the same thing but
```

Comment:
Race is a terrible API we support because it's in ES2015, prefer `some` which is much more informative.